### PR TITLE
Fix NaN/Inf gradient for tf.norm with zero/small inputs

### LINKED ISSUE #12071: Numerical instability of gradient calculation of tf.norm (nan at 0, inf for small values)
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes see below
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac OS X 10.11.6
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: #v1.2.0-5-g435cdfc    1.2.1
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: On CPU
- **GPU model and memory**:
- **Exact command to reproduce**: tf.norm at [0,0] see below for code

```
import numpy as np
import tensorflow as tf
print(tf.GIT_VERSION, "  ", tf.VERSION) #v1.2.0-5-g435cdfc    1.2.1

X = tf.placeholder(tf.float32, shape=(4,None))
Z = tf.norm(X, ord='euclidean', axis=1, name='logit')
var_grad = tf.gradients(Z, [X])

with tf.Session() as sess:
    X_ = np.array([
        [1],  # Grad OK
        [0],  # Grad NaN
        [1e-16],  # Grad OK
        [1e-19] #Grad Inf
    ], dtype=np.float32)
    sess.run(tf.global_variables_initializer())
    print(sess.run((Z, var_grad), feed_dict={X: X_}))
    # Result:
    #(array([9.99999940e-01, 0.00000000e+00, 9.99999951e-17,
    #        0.00000000e+00], dtype=float32), [array([[1.00000012],
    #                                                 [nan],
    #                                                 [1.],
    #                                                 [inf]], dtype=float32)])
```

### Describe the problem
`nan` is calculated for the gradient of `tf.norm` at zero values. For extremely small values `inf` is calculated. Note that the exact result should be 1 in all cases above. 

Above is a minimal example to reproduce it. The problem occurred in a real world scenario, when implementing a custom loss function (the entropy in https://arxiv.org/abs/1611.01449) and two embeddings where too close to each other (distance practically 0).

### Source code / logs
See above 

#### Output of logfile
```
== cat /etc/issue ===============================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64
Mac OS X 10.11.6

== are we in docker =========================================  echo == are we in docker ====================================num echo == are we in docker =========================================  ec==  echo == are we in docker =======================================c++ --version

== uname -a =====================================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy (1.13.0)
protobuf (3.3.0)
tensorflow (1.2.1)

== check for virtualenv ==============  echo == check for virtualenv =====on_b echo == check fo sys  echo == check for virtualenv ============== echo == check for virtualenv ============================================

== cat /etc/issue ===============================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64
Mac OS X 10.11.6

== are we in docker =============================================
No

== compiler =====================================================
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

== uname -a =====================================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy (1.13.0)
protobuf (3.3.0)
tensorflow (1.2.1)

== check for virtualenv =========================================
True

== tensorflow import ============================================
tf.VERSION = 1.2.1
tf.GIT_VERSION = v1.2.0-5-g435cdfc
tf.COMPILER_VERSION = v1.2.0-5-g435cdfc
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
tf_env_collect.sh.txt: line 105: nvidia-smi: command not found

== cuda libs  ===================================================
```

### PR DESCRIPTION
**Pull request by garry00107**
### Description
This PR addresses Issue #12071 where `tf.norm(x)` produces `NaN` gradients when `x` is zero and [Inf](cci:1://file:///Users/garrry/Downloads/Tensorflow/tensorflow/tensorflow/python/ops/math_grad.py:113:0-140:23) gradients when `x` is very small.

The issue arises because the gradient of the Euclidean norm `||x||` involves division by the norm itself: `x / ||x||`. When `||x||` is zero (or very close to it), this results in numerical instability.

### Fix
This change modifies [_EuclideanNormGrad](cci:1://file:///Users/garrry/Downloads/Tensorflow/tensorflow/tensorflow/python/ops/math_grad.py:44:0-72:25) in [tensorflow/python/ops/math_grad.py](cci:7://file:///Users/garrry/Downloads/Tensorflow/tensorflow/tensorflow/python/ops/math_grad.py:0:0-0:0) to handle the zero-norm case explicitly.
- Uses `tf.where` to replace the divisor with 1.0 when the norm is 0, ensuring safe division.
- Masks the final gradient to be 0.0 where the norm was 0, implementing the subgradient convention for the non-differentiable point at the origin.

### Mathematics
At `x=0`, the Euclidean norm is non-differentiable. A common convention in deep learning frameworks (e.g., PyTorch) is to set the subgradient to 0 at this point. This prevents model training from collapsing due to exploding or undefined gradients.

Fixes #12071

### CODE REVIEW

**Comment by garry00107** [comment]
Hi @cantonios  @gbaned , I've updated the PR to address the numerical stability issues. The implementation now uses the standard x * (grad / norm) formula with safety checks for the zero-norm case. Could you please review when you have a moment? Thanks!

**Comment by keerthanakadiri** [comment]
Hi @rohan100jain, Can you please review this PR? Thanks!

**Review by ratgr — CHANGES_REQUESTED** [review]
Great Implementation, I like your solution, but please add UT's(regression tests) to approve your changes

**Comment by garry00107** [comment]
Added regression tests and fixed a keepdims=True broadcast bug found during testing:

Tests added to EuclideanNormGradientTest:
testZeros: updated expected output from [NaN, NaN] → [0.0, 0.0] to document intentional behavior change
testZeroNormSubgradient: 1D all-zero input → zero gradient
testZeroNormSubgradientKeepdims: same with keepdims=True
testZeroNormSubgradient2D: mixed zero/non-zero rows — validates both the fix and existing behavior
testZeroNormGradientNoNanOrInf: explicit NaN/Inf guard
Additional bug fix in math_grad.py: When keep_dims=True, the zero-condition had shape (1,1) while grad_input had shape (1,2), causing where_v2 to produce NaN. Fixed by broadcasting is_zero to grad_input shape before the final masking step.

All 20 EuclideanNormGradientTest tests pass (graph + eager mode, float32 + float64).
